primitives/src/lib.rs

```diff
@@ -272,10 +272,61 @@ pub struct PalletName(pub Vec<u8>);
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
 pub struct DispatchableName(pub Vec<u8>);
 
+/// Create a `Version` struct with an upper limit.
+#[macro_export]
+macro_rules! storage_migration_ver {
+    ($ver:literal) => {
+        #[derive(Encode, Decode, Clone, Copy, PartialEq, Eq, PartialOrd, Ord)]
+        pub struct Version(u8);
+
+        impl Version {
+            const MAX: u8 = $ver;
+
+            const fn new(ver: u8) -> Option<Self> {
+                if ver <= Self::MAX {
+                    Some(Self(ver))
+                } else {
+                    None
+                }
+            }
+        }
+
+        impl Default for Version {
+            fn default() -> Self {
+                Version(0)
+            }
+        }
+
+        impl sp_std::convert::TryFrom<u8> for Version {
+            type Error = &'static str;
+
+            fn try_from(ver: u8) -> Result<Self, Self::Error> {
+                Self::new(ver).ok_or("Unsupported version")
+            }
+        }
+    };
+}
+
+/// Helper macro which execute the `$body` if `$curr` is less than version `$ver`.
+/// It also updates `<StorageVersion>` in the current pallet to `$ver`.
+#[macro_export]
+macro_rules! storage_migrate_on {
+    ($curr: expr, $ver:literal, $body: block) => {{
+        const TARGET_VERSION: Version = Version::new($ver).unwrap();
```

Comment:
This is *really* nice. We'll get immediate linting errors and don't have to wait until runtime. Great work.